core/src/main/java/com/linecorp/armeria/common/ContextStorage.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+import static com.linecorp.armeria.internal.common.RequestContextUtil.defaultContextStorage;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.util.UnstableApi;
+
+/**
+ * The storage for storing {@link RequestContext}.
+ *
+ * <p>If you want to implement your own storage or add some hooks when a {@link RequestContext} is pushed
+ * and popped, you should use {@link ContextStorageProvider} or {@link Flags#contextStorage()}.
+ * Here's an example that sets MDC before {@link RequestContext} is pushed:
+ *
+ * <pre>{@code
+ * > public class MyStorage implements ContextStorageProvider {
+ * >     @Override
+ * >     public ContextStorage newContextStorage() {
+ * >         ContextStorage storage = ContextStorage.ofDefault();
+ * >         return new ContextStorage() {
+ * >
+ * >             @Nullable
+ * >             @Override
+ * >             @SuppressWarnings("unchecked")
+ * >             public <T extends RequestContext> T push(RequestContext toPush) {
+ * >                 setMDC(...); // using toPush
+ * >                 return storage.push(toPush);
+ * >             }
+ * >
+ * >             @Override
+ * >             public void pop(RequestContext current, @Nullable RequestContext toRestore) {
+ * >                 if (toRestore != null) {
+ * >                     setMDC(...); // using toRestore
+ * >                 }
+ * >                 storage.pop(current, toRestore);
+ * >             }
+ * >             ...
+ * >      }
+ * > }
+ * }</pre>
+ */
+@UnstableApi
+public interface ContextStorage {
+
+    /**
+     * Returns the default {@link ContextStorage} which stores the {@link RequestContext} in the thread-local.
+     */
+    static ContextStorage ofDefault() {
```

Comment:
Maybe `threadLocal()`, because the default storage changes based on flag/SPI?

Reply:
That's a good point. Let me make a change. Thanks!